cpp/src/buffer/buffer.cpp

```diff
@@ -139,8 +155,179 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
     );
 }
 
+std::unique_ptr<Buffer> Buffer::copy_slice(
+    std::ptrdiff_t offset,
+    std::size_t length,
+    MemoryReservation& target_reserv,
+    rmm::cuda_stream_view stream
+) const {
+    RAPIDSMPF_EXPECTS(
+        target_reserv.size() >= length, "reservation is too small", std::invalid_argument
+    );
+    RAPIDSMPF_EXPECTS(
+        offset >= 0 && offset <= std::ptrdiff_t(size),
+        "offset can't be greater than size",
+        std::invalid_argument
+    );
+    RAPIDSMPF_EXPECTS(
+        offset + std::ptrdiff_t(length) <= std::ptrdiff_t(size),
+        "offset + length can't be greater than size",
+        std::invalid_argument
+    );
+
+    // lambda to allocate a buffer and perform a memcpy
+    auto do_alloc_and_cuda_memcpy_async = [&](cudaMemcpyKind kind,
+                                              cuda::std::uint8_t* source) {
+        // allocate buffer using the target reservation
+        auto out_buf = target_reserv.br()->allocate(
+            target_reserv.mem_type(), length, stream, target_reserv
+        );
+
+        if (length > 0) {
+            // if this buffer has an event, ask the current stream to wait for it, before
+            // performing the memcpy
+            if (event_) {
+                RAPIDSMPF_CUDA_TRY(cudaStreamWaitEvent(stream, event_->event()));
+            }
```

Comment:
Now that `cudaMemcpyAsync` is used, do we still need to wait?

Reply:
@madsbk Based on what @wence- suggested, this would be more of a marker put on the `stream` to wait on the event asynchronously. This would be more relevant if the `event` was recorded on a different stream than `stream`.